### PR TITLE
Move Person to ComponentModel, make property changes Asynchronous

### PR DESCRIPTION
**Pull request by nevercast**
Blocked by: https://github.com/RaySingerNZ/cpp/tree/patch-code2
_patch-code2 needs to be pulled BEFORE this PR_

Lots of changes here. Thought I'd play around.

Person now implements ComponentModel's INotifyPropertyChanged which seems to be the Microsoft way of notifying that a property value has changed. This is used in Data binding.

I've extended the PropertyChanged event with the PropertyValueChangedEventArgs class that holds the Old and New value properties. This replaces the ValueChanged EventArgs/Delegate pair that existed before.

The events are also dispatched as Tasks using the TaskScheduler. Delegate invocations don't support tasks directly but do support the Async pattern from earlier .NET versions (Using BeginInvoke and EndInvoke). Using TaskFactory.FromAsync this is encapsulated as a Task. A method built to wrap the old async pattern in to the Task pattern. 

I've tidied a bit of the code, and covered it in documentation.

I also implemented GetHashCode for using Person in collections.

### CODE REVIEW

**Comment by nevercast** [comment]
Pulling this will pull Ray's changes anyway since its an ancestor. 

@raysingernz Do you still wish to PR your patch-code2 branch?

**Comment by RaySingerNZ** [comment]
Thanks @nevercast that works with me.

File: DelegatesAndEventsDemo/Person.cs [hunk 1]
```diff
@@ -1,87 +1,163 @@
...
-                        args.NewValue = firstName;
-                        ValueDelegate(this, args);
-                    }
+                    // Because the changes are invocated by the TaskScheduler, we get race conditions
+                    // We need to set the properties before calling the event
+                    // Otherwise calls to ToString will show wrong values, but more strangely sometimes
+                    // they will show correct values. Depending on when things are called
+                    var old = this.firstName;
+                    this.firstName = value;
+                    OnPropertyValueChanged("FirstName", old, value);
```
Comment (nateforsyth) [inline DelegatesAndEventsDemo/Person.cs:55]: I'm not following what was done here, how have you replaced all of the event handling code with this one Class? Is it merely further specialisation deriving from the EventArgs class?

**Comment by nevercast** [comment]
I'm going to close this :+1:

**Pull request closed without merging**